### PR TITLE
Side-channel proofing PKCS#1 1.5 paths (CVE-2023-6258)

### PR DESCRIPTION
**Pull request by simo5**
Fixes for CVE-2023-6258 (Marvin)

### CODE REVIEW

File: src/asymmetric_cipher.c [hunk 1]
```diff
@@ -296,12 +296,28 @@ static int p11prov_rsaenc_decrypt(void *ctx, unsigned char *out, size_t *outlen,
...
+    /* Special handling against PKCS#1 1.5 side channel leaking */
+    if (mechanism.mechanism == CKM_RSA_PKCS) {
+        ret = side_channel_free_Decrypt(encctx->provctx, sess, (void *)in,
+                                        inlen, out, &out_size);
+        /* the error case need to be handled in a side-channel free way, so
+         * conditionals ineed to be idempotent. Setting outlen unconditionally
+         * is fine because out_size is initialized to the value of outlen
+         * itself, and the value should not matter in an error condition */
+        *outlen = out_size;
+        if (ret == CKR_OK) {
```
Comment (tomato42) [inline src/asymmetric_cipher.c:308]: this will leak, the compare, the subsequent jump, and the secret data dependent load of constant from memory

Comment (simo5) [reply src/asymmetric_cipher.c]: I added constant time functions for this.

**Review by tomato42 — APPROVED** [review]

**Comment by tomato42** [comment]
Just to be explicit: the "approved" status is because they _look_ fine, not because I've verified to _be_ side-channel free.

**Comment by tomato42** [comment]
this issue was assigned CVE-2023-6258

**Comment by simo5** [comment]
Merging this now that debian fixes are in and all tests pass, macos softoken is still broken because the CI still uses the  broken NSS versions and the softhsm for other reasons that seem also macports related.

**Pull request merged**